services/vmagent-config-updater/main.go

```diff
@@ -1,56 +1,129 @@
 	"fmt"
 	"log"
 	"math/rand"
 	"net/http"
+	"os"
+	"strings"
 	"sync"
 	"time"
 
-	"gopkg.in/yaml.v2"
+	"gopkg.in/yaml.v3"
 )
 
+func parseFlagValue[T cmp.Ordered](v string, defaultValue T) (any, error) {
+	var err error
+	value := defaultValue
+	if len(v) > 0 {
+		if _, ok := any(value).(time.Duration); ok {
+			return time.ParseDuration(v)
+		}
+		_, err = fmt.Sscanf(v, "%v", &value)
+	}
+	return value, err
+}
+
+type arrayFlag[T cmp.Ordered] struct {
+	values       []T
+	defaultValue T
+}
+
+func (af *arrayFlag[T]) String() string {
+	if len(af.values) > 0 {
+		strVals := make([]string, len(af.values))
+		for i, v := range af.values {
+			strVals[i] = fmt.Sprintf("%v", v)
+		}
+		return strings.Join(strVals, ",")
+	}
+	return fmt.Sprintf("%v", af.defaultValue)
+}
+
+func (af *arrayFlag[T]) Set(value string) error {
+	for _, v := range strings.Split(value, ",") {
+		if val, err := parseFlagValue(v, af.defaultValue); err != nil {
+			return fmt.Errorf("failed to parse value %v for type %t", val)
+		} else {
+			af.values = append(af.values, val.(T))
+		}
+	}
+	return nil
+}
+
+func (af *arrayFlag[T]) total() []T {
+	if len(af.values) == 0 {
+		return []T{af.defaultValue}
+	}
+	return af.values
+}
+
+func (af *arrayFlag[T]) getArg(idx int) T {
+	if len(af.values) == 0 || idx >= len(af.values) {
+		return af.defaultValue
+	}
+	return af.values[idx]
+}
+
+func newArrayFlag[T cmp.Ordered](name string, defaultValue T, description string) *arrayFlag[T] {
+	description += "\nSupports an `array` of values separated by comma or specified via multiple flags."
+	a := &arrayFlag[T]{
+		defaultValue: defaultValue,
+	}
+	flag.Var(a, name, description)
+	return a
+}
+
 var (
 	listenAddr                 = flag.String("httpListenAddr", ":8436", "TCP address for incoming HTTP requests")
-	targetsCount               = flag.Int("targetsCount", 100, "The number of scrape targets to return from -httpListenAddr. Each target has the same address defined by -targetAddr")
-	targetAddr                 = flag.String("targetAddr", "demo.robustperception.io:9090", "Address with port to use as target address the scrape config returned from -httpListenAddr")
-	scrapeInterval             = flag.Duration("scrapeInterval", time.Second*5, "The scrape_interval to set at the scrape config returned from -httpListenAddr")
-	scrapeConfigUpdateInterval = flag.Duration("scrapeConfigUpdateInterval", time.Minute*10, "The -scrapeConfigUpdatePercent scrape targets are updated in the scrape config returned from -httpListenAddr every -scrapeConfigUpdateInterval")
-	scrapeConfigUpdatePercent  = flag.Float64("scrapeConfigUpdatePercent", 1, "The -scrapeConfigUpdatePercent scrape targets are updated in the scrape config returned from -httpListenAddr ever -scrapeConfigUpdateInterval")
+	labelName                  = newArrayFlag("labelName", "instance", "Label name, which differs for all state copies")
+	jobName                    = newArrayFlag("jobName", "node_exporter", "Scrape job name")
```

Comment:
Shall we explicitly set these flags in the vmagent chart? This would demonstrate how config reloader is used for specific case

Reply:
okay

Reply:
do you mean all array flags?

Reply:
I was thinking about this https://github.com/VictoriaMetrics/prometheus-benchmark/blob/0a5d35cf34e1f909e20c50a93c33f7e0ba38cca8/chart/templates/vmagent/deployment.yaml#L32-L40

Reply:
ping